Sources/Crypto/Cipher/AuthenticatedCipher.swift

```diff
@@ -0,0 +1,132 @@
+        var buffer = Data()
+
+        try reset(key: key, iv: iv, mode: .encrypt)
+        try update(data: data, into: &buffer)
+        try finish(into: &buffer)
+
+        let tag = try gcmTag()
+
+        return (buffer, tag)
+    }
+
+    /// Decrypts the supplied ciphertext back to plaintext. This method will call `reset(key:iv:mode:)`, `update(data:into:)`,
+    /// and `finish(into:)` automatically.
+    ///
+    ///     let key: Data // 32-bytes
+    ///     let iv: Data // 12-bytes
+    ///     let (ciphertext, tag) = try AES256GCM.encrypt("vapor", key: key, iv: iv)
+    ///     try AES256GCM.decrypt(ciphertext, key: key, iv: iv, tag: tag) // "vapor"
+    ///
+    /// - parameters:
+    ///     - data: Ciphertext data to decrypt.
+    ///     - key: Cipher key to use for decryption.
+    ///            This key must be an appropriate length for the cipher you are using. See `CipherAlgorithm.keySize`.
+    ///     - iv: Initialization vector to use for decryption.
+    ///           The IV must be an appropriate length for the cipher you are using. See `CipherAlgorithm.ivSize`.
+    ///     - tag: Authentication tag for GCM-mode ciphers.
+    ///           The tag must be 16 bytes
+    /// - returns: Decrypted plaintext.
+    /// - throws: `CryptoError` if reset, update, or finalization steps fail or if data conversion fails.
+    public func decrypt(_ data: LosslessDataConvertible, key: LosslessDataConvertible, iv: LosslessDataConvertible, tag: LosslessDataConvertible) throws -> Data {
+        var buffer = Data()
+
+        try reset(key: key, iv: iv, mode: .decrypt)
+        try update(data: data, into: &buffer)
+        try gcmTag(tag.convertToData())
+        try finish(into: &buffer)
+
+        return buffer
+    }
+
+    /// Gets the GCM Tag from the CIPHER_CTX struct.
+    ///
+    /// Note: This _must_ be called after `finish()` to retrieve the generated tag.
+    ///
+    /// - throws: `CryptoError` if tag retrieval fails
+    public func gcmTag() throws -> Data {
+        var buffer = Data(count: AuthenticatedCipher.gcmTagLength)
+
+        guard buffer.withMutableByteBuffer({ EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_GET_TAG, Int32(AuthenticatedCipher.gcmTagLength), $0.baseAddress!) }) == 1 else {
+            throw CryptoError.openssl(identifier: "EVP_CIPHER_CTX_ctrl", reason: "Failed getting tag (EVP_CTRL_CCM_GET_TAG).")
+        }
+
+        return buffer
+    }
+
+    /// Sets the GCM Tag in the CIPHER_CTX struct.
+    ///
+    /// Note: This _must_ be called before `finish()` to set the tag.
+    ///
+    /// - throws: `CryptoError` if tag set fails
+    public func gcmTag(_ tag: Data) throws {
+        var buffer = tag
+
+        guard buffer.withMutableByteBuffer({ EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_TAG, Int32(AuthenticatedCipher.gcmTagLength), $0.baseAddress!) }) == 1 else {
```

Comment:
I think this will have unexpected results for a tag argument _shorter_ than `AuthenticatedCipher.gcmTagLength`. I would suggest checking for the exactly correct tag length here, and maybe adding a test that verifies that using a too short tag does not crash.